azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/AbstractAzureMojo.java

```diff
@@ -272,74 +274,97 @@ public int getHttpProxyPort() {
         return httpProxyPort;
     }
 
-    public Azure getAzureClient() throws AzureAuthFailureException {
+    public Azure getAzureClient() throws AzureAuthFailureException, AzureExecutionException {
         if (azure == null) {
             if (this.authentication != null && (this.authentication.getFile() != null || StringUtils.isNotBlank(authentication.getServerId()))) {
                 // TODO: remove the old way of authentication
                 Log.warn("You are using an old way of authentication which will be deprecated in future versions, please change your configurations.");
                 azure = new AzureAuthHelperLegacy(this).getAzureClient();
             } else {
-                initAuth();
-                azure = getAzureClientByAuthType();
+                azure = createAzureClient();
             }
             if (azure == null) {
                 getTelemetryProxy().trackEvent(INIT_FAILURE);
                 throw new AzureAuthFailureException(AZURE_INIT_FAIL);
             }
             printCurrentSubscription(azure);
-            getTelemetryProxy().addDefaultProperty(AUTH_TYPE, authType);
+            getTelemetryProxy().addDefaultProperty(AUTH_TYPE, getAuthType());
             getTelemetryProxy().addDefaultProperty(AUTH_METHOD, getAuthMethod());
             // Repopulate subscriptionId in case it is not configured.
             getTelemetryProxy().addDefaultProperty(SUBSCRIPTION_ID_KEY, azure.subscriptionId());
-
         }
         return azure;
     }
 
-    protected Azure getAzureClientByAuthType() throws AzureAuthFailureException {
+    protected String getAuthType() {
+        return StringUtils.firstNonBlank(auth == null ? null : auth.getAuthType(), authType);
+    }
+
+    protected Subscription selectSubscription(Azure az, TextIO textIO, Subscription[]subscriptions) throws AzureExecutionException {
```

Comment:
I think the signature `selectSubscription(Subscription[] subs, Subscription current)` is more natural and straight.

seems `az` only used to get current subscription. and 
IMO, `textIO` should be static or singleton?

Reply:
because when Azure instance is created, there is no way of changing defaultSubscription,s so we need to wrap it